iAPI Docs: Core concepts - How it works scheme overview

WPLake opened the pull request:
This guide is part of the [new Core Concepts section](https://github.com/WordPress/gutenberg/issues/62921) of the Interactivity API documentation. It aims to showcase and explain the entire API's workflow based on a single page request. This should help devs create and structure a general process roadmap, which other guides will then detail.

WPLake committed:
iAPI Docs: Core concepts - How it works scheme overview

WPLake committed:
iAPI Docs: Core concepts - How it works scheme overview

WPLake committed:
iAPI Docs: Core concepts - How it works scheme overview

WPLake committed:
iAPI Docs: Core concepts - How it works scheme overview

juanmaguitar commented:
@WPLake Thanks for submitting this PR. This content is really great to help understand iAPI!
I think it will be a great addition to the "Core Concepts" section.

There are a few things I consider could be added to the post to make it even more impactful:
- It would be great if you could add links to other parts of the docs. For example, when you mention directives, it would be great to provide a link to the part of the docs that explains that directive
- I think this example would be even more complete if some interaction is added - we're talking about the Interactivity API, right? :). As the example is about an accordion, adding a button to show/hide some elements in it would be great. 
You probably already have this on your radar, but I wanted to mention it just in case.
- I think the title needs to be more descriptive of the content. Some suggestions I have are: 
  - Backend and Client-Side Processing: A detailed guide
  - How the Interactivity API works: An In-Depth Look
  - A Detailed Walkthrough of Backend and Client-Side Processing

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
```diff
@@ -0,0 +1,219 @@
+To get a complete picture of how the API works, let's review a request to the imagined `/interactive` page.
```
```suggestion
To get a complete picture of how the Interactivity API works, let's review a request to an imagined `/interactive` page.
```

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
```diff
@@ -0,0 +1,219 @@
+To get a complete picture of how the API works, let's review a request to the imagined `/interactive` page.
+
+The `/interactive` page is a plain page that contains an interactive "accordion" block among static content. In this
```
```suggestion
The `/interactive` page is a standard page that contains an accordion block among static content. In this
```

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
```diff
@@ -0,0 +1,219 @@
+The `/interactive` page is a plain page that contains an interactive "accordion" block among static content. In this
+case, the request will pass through the following phases:
+
+## Part A: Backend
```
```suggestion
## Server side processing begins
```

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
```diff
@@ -0,0 +1,219 @@
+
+The initial request to the `/interactive` page is processed by WordPress, and all the hooks are fired as usual.
+
+### 2. Page body HTML generation
```
```suggestion
### 2. HTML DOM processing
```

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
````diff
@@ -0,0 +1,219 @@
+accordion block
+is reached:
+
+```php
````
```suggestion
```php
// render.php
```

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
```diff
@@ -0,0 +1,219 @@
+
+The following will happen:
+
+#### 2.1) The `wp_interactivity_state` call defines the state variables of our block.
```
```suggestion
#### 2.1) The `wp_interactivity_state` call initializes some server-side state for the block.
```

colorful-tones commented on docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md:
````diff
@@ -0,0 +1,219 @@
+```
+
+Note: The `wp_interactivity_state` and `wp_interactivity_data_wp_context` calls are optional,
+while `data-wp-interactive="accordion"` is a necessary directive that defines our block.
````
```suggestion
while `data-wp-interactive="accordion"` is a required directive to tell the Interactivity API that we want this element and its nested DOM elements to be recognized by the Interactivity API.
```

colorful-tones reviewed (commented):
I love these explanations and think they'll be super helpful. Thanks for writing this all up @WPLake 

I started to make some suggestions. I agree with @juanmaguitar that we should try to link to API functions when possible.

WPLake committed:
Apply suggestions from code review

Co-authored-by: Damon Cook <colorful-tones@users.noreply.github.com>

WPLake committed:
Update docs/reference-guides/interactivity-api/core-concepts/how-it-works-scheme-overview.md

Co-authored-by: Damon Cook <colorful-tones@users.noreply.github.com>

WPLake committed:
Apply suggestions from code review

Co-authored-by: Damon Cook <colorful-tones@users.noreply.github.com>

WPLake committed:
Update and rename how-it-works-scheme-overview.md to how-the-interactivity-api-works

WPLake commented:
@juanmaguitar Glad to hear you like the suggested page!

1. Setup the suggested title:
We have added the title "How the Interactivity API Works: An In-Depth Look" and renamed the file to reflect this change.

2. Added remaining elements to the accordion example:
The initial intention was to keep the example simple, given that it is a guide for explaining the scheme. However, with the upcoming Live Playground, the example should now be complete.

3. Links to API Functions:
Could you advise where to find the appropriate URLs for the API functions? If the final URLs for the new section are not yet available, should we include placeholder links for now?

WPLake commented:
@colorful-tones That’s great to hear! Thank you for the suggestions; we've applied them.

WPLake committed:
Update how-the-interactivity-api-works

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,276 @@
+# How the Interactivity API works: An In-Depth Look
+
+To get a complete picture of how the Interactivity API works, let's review a request to an imagined `/interactive` page.
```
This is a nitpick, but I was stumped for a second by what was mentioned by "imagined".

```suggestion
To get a complete picture of how the Interactivity API works, let's review a request to a hypothetical `/interactive` page.
```

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,276 @@
+
+#### 3.2) WP automatically enqueues the block script (only for Gutenberg blocks).
+
+The Interactivity API blocks have their own JavaScript, which defines actions, and this JavaScript is automatically
```
There's no such thing as "Interactivity API blocks" and I think it's confusing to say that they "have their own JavaScript". 

It would be clearer to say "Blocks using the Interactivity API can define state and actions in JavaScript files. This JS is automatically enqueued by WordPress.

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
````diff
@@ -0,0 +1,276 @@
+```
+
+If the block has no JS code, the interactivity JS won't be enqueued. It may look confusing, but after considering it,
+it becomes clear that no block JS means no actions and no client-side states, so the server will deliver the final
````
```suggestion
it becomes clear that no block JS means no actions and no client-side state, so the server will deliver the final
```

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,276 @@
+
+If the block has no JS code, the interactivity JS won't be enqueued. It may look confusing, but after considering it,
+it becomes clear that no block JS means no actions and no client-side states, so the server will deliver the final
+markup, which won't have any changes down the line.
```
What is meant by "no changes down the line"? I find the phrasing confusing.

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
````diff
@@ -0,0 +1,276 @@
+</div> 
+```
+
+Note: The `wp_interactivity_state` and `wp_interactivity_data_wp_context` calls are optional,
````
It should be made clear that `wp_interactivity_state` and `wp_interactivity_data_wp_context` are NOT optional in this example. The Accordion would not work without them. 

They are only "optional" in the sense that they are not required for the Interactivity API to work.

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,276 @@
+#### 2.3) WordPress, upon encountering `data-wp-interactive="accordion"`, processes the block directives.
+
+In this case, there are 7 directives to be processed: 2 `data-wp-bind--hidden`, 5 `data-wp-text`
+and `data-wp-class--dark`. Using the HTML
```
Can we add a hyperlink to the HTML API here?

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,276 @@
+process them and change the markup.
+
+As a result, the span will contain the translated text and the hidden attribute
+will be added to the matching (first) item. The `data-wp-class--dark` directive will be skipped, as no such state is
```
It's not correct to say that the "directive will be skipped". The directive will be processed but its value is unset so the `dark` class will not be applied.

michalczaplinski commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,276 @@
+#### 3.1) WP automatically adds the `interactivity.min.js` script alias.
+
+Besides adding the `<script>` tags for all the
+assets used on the page, WP will automatically add the `@wordpress/interactivity` alias, pointing to
```
It should be clarified what this "alias" is and where it comes from. It's an entry in the [import map](https://make.wordpress.org/core/2024/03/04/script-modules-in-6-5/) that is added by WordPress

michalczaplinski reviewed (changes requested):
I like the idea of the "how it works" document, but would be great to improve further:

1. Let's shorten and simplify the code examples. The vast majority of the code is not referred to in the text. For example, the HTML markup of the accordion block has many CSS classes, the `store` has multiple getters, that serve no purpose in the explanation, etc.

2. The phrase "state variable" or "state variables"  is not something that we use in the Interactivity API documentation. It should be either "state", "actions" or "store" (to refer to the whole reactive object that contains the state, actions, callbacks, etc.)

3. Could we run the text through Grammarly, an LLM or similar? I notice a few places where the grammar is less than perfect and I'm afraid it hinders understanding.

4. The last section `Part 3: How to "enable" the Interactivity API` seems unrelated to the first two. Could it be in a separate document?

WPLake committed:
Apply suggestions from code review

Co-authored-by: Michal <mmczaplinski@gmail.com>

WPLake committed:
Update how-the-interactivity-api-works

WPLake commented:
@michalczaplinski Glad to hear you like the idea. 

1. **"Let's shorten and simplify the code examples"**  
   Initially, the examples highlighted only the target parts, making them clearer but incomplete. JuanMa Garrido suggested completing the examples, which we did. In my view, it makes sense to keep the whole example in place, as the [code snippets will be interactive](https://github.com/WordPress/gutenberg/issues/62921#issuecomment-2268383580) in the future.

2. **The phrase "state variable" or "state variables"**  
   Changed as needed.

3. **Reviewed through Grammarly, an LLM, or similar**  
   Applied for improvements.

4. **The last section, Part 3: How to "enable" the Interactivity API**  
   Removed.

We've also addressed your other comments as well.

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+# How the Interactivity API works: An In-Depth Look
```
This file should be Markdown.

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+
+### 2. HTML DOM processing
+
+At this stage, the current theme generates the markup for the request. During this process, the section containing the interactive accordion block is rendered:
```
Is this a block? Is it a classic theme?

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+
+The following will happen:
+
+#### 2.1) The `wp_interactivity_state` call initializes some server-side state for the block.
```
We are now referring to the state as global state to differentiate it from context, which is now called local context. More information [here](https://developer.wordpress.org/block-editor/reference-guides/interactivity-api/core-concepts/undestanding-global-state-local-context-and-derived-state/).

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
````diff
@@ -0,0 +1,193 @@
+data-wp-context='{"openItemMessage":"Current open item is","closedItemMessage":"Items are closed","lastOpenItemMessage":"Last opened item is"}
+```
+
+#### 2.3) When WordPress encounters `data-wp-interactive="accordion"`, it processes the block directives.
````
Is this referring to the server or the client? If it's on the server, it still doesn't work that way; only blocks marked as interactive in the block.json are processed, or the HTML passed to the `wp_interactivity_process_directives` function.

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+#### 2.3) When WordPress encounters `data-wp-interactive="accordion"`, it processes the block directives.
+
+In this case, there are 7 directives to be processed: 2 `data-wp-bind--hidden`, 5 `data-wp-text`, and `data-wp-class--dark`. 
+Using the [HTML API](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/) and the defined data (the state from the global block-related PHP variable and variables as JSON attributes), WordPress processes these directives and updates the markup.
```
Here you mention a block, but it's not clear from the beginning whether you're talking about blocks or classic themes. Or I don't understand what you mean by global block related PHP variable and variables as JSON attributes.

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+Using the [HTML API](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/) and the defined data (the state from the global block-related PHP variable and variables as JSON attributes), WordPress processes these directives and updates the markup.
+
+As a result, the `<span>` will contain the translated text, and the `hidden` attribute will be added to the matching (first) item. 
+The `data-wp-class--dark` directive will be processed but its value is unset so the dark class will not be applied. This directive represents a client-only state, which will be defined on the client side.
```
> This directive represents a client-only state, which will be defined on the client side.

Do you mean that the `data-wp-class` directive is client-side only?

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+#### 3.1) WordPress automatically adds the `interactivity.min.js` script alias.
+
+In addition to adding the `<script>` tags for all assets used on the page, WordPress will automatically include the `@wordpress/interactivity` alias, 
+which points to the `/wp-includes/js/dist/interactivity.min.js` script. This occurs only on pages that contain at least one interactive block.
```
> This occurs only on pages that contain at least one interactive block.

This is usually true when developing with `wp-scripts`, but we should be more careful in how we say this to manage expectations, as the script module `@wordpress/interactivity` is only added to the page when another script module that has it as a dependency is enqueued.

luisherranz commented on docs/reference-guides/interactivity-api/core-concepts/how-the-interactivity-api-works:
```diff
@@ -0,0 +1,193 @@
+
+### 3. Scripts enqueuing
+
+#### 3.1) WordPress automatically adds the `interactivity.min.js` script alias.
```
We should be talking about script modules to make it clear that they are not scripts.

luisherranz reviewed (changes requested):
_I have reviewed only the first part. I'll finish later._

Overall, I think it's okay to add an entry for "How it works", but I believe it needs more work and we need to carefully consider at what level we should provide the explanation and then think more carefully about what are the different steps involved and what things need to be explained and which ones should be assumed to be known by the user.
